dafny-lang/dafny
# Add replaceable modules

### Description
- Move some C# code from `Dafny.atg` and `Parser.frame` to a new file named `ParserNonGeneratedPart.cs`, so C# IDEs can provide tooling for this code.

- Enable the following syntax
```dafny
replaceable module FooModule {
  method Foo() returns (i: int) 
    ensures i >= 2
}

module Nesting {
  module ConcreteFoo replaces FooModule {
    method Foo() returns (i: int) {
      return 3;
    }
  }
}

method Main() {
  var x := FooModule.Foo();
  print x; // Prints 3
}
```

Check out one the test-cases for an example of how you define a multi-target application using this feature: https://github.com/dafny-lang/dafny/blob/66c1b1ee414bb4a5060507dbb0dd333697968e02/Source/IntegrationTests/TestFiles/LitTests/LitTest/comp/placeholders/complex/user.dfy

### Caveats

Refining the `replaceable ` module is not supported, and it will not be supported for some time. There's a conflict between how refinement works and how replaceable modules work. 

Replaceable modules are implemented as a post-resolution feature. It swaps out references to the replaceable with the replacement. It has to work like this because `dafny verify` should work without the replacement being part of the source, so refinement can not rely on the replacement module being present.

Refinement, as it is currently implemented, copies non-abstract parts of the refinee into the refiner during resolution, so you would need access to the replacement module during resolution, if you want to further refine it. 

We could change refinement so it does use the replacement module _if_ its available, to influence refinement, but I consider that detrimental to maintainability so I think we should not do this. Refinement being implemented using copying is already problematic because it increases compilation time and generated code size. I think we should change refinement so it uses references to the refinee's members instead of copying them, and then we can also implement refinement of placeholders.

### How has this been tested?
- Add a happy flow test containing:
  - A placeholder module that is replaced
  - A placeholder module that refines an abstract module, and is replaced
  - A placeholder module that replaces another placeholder module, and is replaced
  - Various abstract imports using placeholders
- Add complex tests that use different back-ends for a common library, showing off the main driver behind the feature.
- Add a test that exposes all added errors
  - A placeholder module can not be refined
  - A placeholder module was not replaced
  - A placeholder module was replaced more than once
  - Cycle of replacement

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>
